pkg/optimize/acl/reduceCubes.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package acloptimizer
+
+import (
+	"github.com/np-guard/models/pkg/ds"
+	"github.com/np-guard/models/pkg/netset"
+)
+
+func reduceACLCubes(aclCubes *aclCubesPerProtocol) {
+	allTCP := allTCPUDP(aclCubes.tcpAllow)
+	allUDP := allTCPUDP(aclCubes.udpAllow)
+	allicmp := allICMP(aclCubes.icmpAllow)
+	aclCubes.anyProtocolAllow = allTCP.Intersect(allUDP).Intersect(allicmp)
+	subtractAnyProtocolCubes(aclCubes)
+}
+
+func allTCPUDP(tcpudpAllow ds.TripleSet[*netset.IPBlock, *netset.IPBlock, *netset.TCPUDPSet]) *srcDstProductLeft {
+	res := ds.NewProductLeft[*netset.IPBlock, *netset.IPBlock]()
+	for _, p := range tcpudpAllow.Partitions() {
+		if p.S3.Equal(netset.NewAllTCPOnlySet()) || p.S3.Equal(netset.NewAllUDPOnlySet()) { // all tcp or udp ports
```

Comment:
call `netset.NewAllTCPOnlySet()` and `netset.NewAllUDPOnlySet()` only once outside of the loop